### PR TITLE
[kubeadm/app/]switch to github.com/pkg/errors

### PR DESCRIPTION
**Pull request by yuexiao-wang**
Signed-off-by: yuexiao-wang <wang.yuexiao@zte.com.cn>

**What type of PR is this?**
> Uncomment only one, leave it on its own line:
>
> /kind api-change
> /kind bug

 /kind cleanup

> /kind design
> /kind documentation
> /kind failing-test
> /kind feature
> /kind flake

**What this PR does / why we need it**:
switch to github.com/pkg/errors the packages as follows:
cmd/kubeadm/app/cmd/upgrade
cmd/kubeadm/app/cmd/util
cmd/kubeadm/app/discovery
cmd/kubeadm/app/discovery/file
cmd/kubeadm/app/phases/addons/dns
cmd/kubeadm/app/phases/addons/proxy
cmd/kubeadm/app/phases/bootstraptoken/clusterinfo/
cmd/kubeadm/app/phases/etcd

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes # Fixes # kubernetes/kubeadm#1183

**Special notes for your reviewer**:
@fabriziopandini  @neolit123 

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

### CODE REVIEW

**Review by neolit123 — APPROVED** [review]
/lgtm
/kind cleanup

@yuexiao-wang thanks, might have been better to change all kubeadm packages in a single PR.

File: cmd/kubeadm/app/cmd/upgrade/apply.go [hunk 1]
```diff
@@ -250,13 +251,13 @@ func EnforceVersionPolicies(flags *applyFlags, versionGetter upgrade.VersionGett
 	if versionSkewErrs != nil {
 
 		if len(versionSkewErrs.Mandatory) > 0 {
-			return fmt.Errorf("The --version argument is invalid due to these fatal errors:\n\n%v\nPlease fix the misalignments highlighted above and try upgrading again", kubeadmutil.FormatErrMsg(versionSkewErrs.Mandatory))
+			return fmt.Errorf("the --version argument is invalid due to these fatal errors:\n\n%v\nPlease fix the misalignments highlighted above and try upgrading again", kubeadmutil.FormatErrMsg(versionSkewErrs.Mandatory))
```
Comment (fabriziopandini) [inline cmd/kubeadm/app/cmd/upgrade/apply.go]: Errors.new?

File: cmd/kubeadm/app/cmd/upgrade/apply.go [hunk 2]
```diff
@@ -250,13 +251,13 @@ func EnforceVersionPolicies(flags *applyFlags, versionGetter upgrade.VersionGett
...
 		}
 
 		if len(versionSkewErrs.Skippable) > 0 {
 			// Return the error if the user hasn't specified the --force flag
 			if !flags.force {
-				return fmt.Errorf("The --version argument is invalid due to these errors:\n\n%v\nCan be bypassed if you pass the --force flag", kubeadmutil.FormatErrMsg(versionSkewErrs.Skippable))
+				return fmt.Errorf("the --version argument is invalid due to these errors:\n\n%v\nCan be bypassed if you pass the --force flag", kubeadmutil.FormatErrMsg(versionSkewErrs.Skippable))
```
Comment (fabriziopandini) [inline cmd/kubeadm/app/cmd/upgrade/apply.go]: errors.Errorf
(Same applies to all fmt.Errorf)

**Review by fabriziopandini — CHANGES_REQUESTED** [review]
@yuexiao-wang thanks!
Only one nit and then ok for approval

File: cmd/kubeadm/app/cmd/upgrade/apply.go (hunk 2 above)
Comment (yuexiao-wang) [reply cmd/kubeadm/app/cmd/upgrade/apply.go]: Thanks. Got it and  I need to append these applications to my PR.
There are so many fmt.Errorf in kubeadm packages. so I only switch to errors.Wrap and error.New in my PR.

File: cmd/kubeadm/app/cmd/upgrade/apply.go (hunk 1 above)
Comment (yuexiao-wang) [reply cmd/kubeadm/app/cmd/upgrade/apply.go]: Got it

**Comment by yuexiao-wang** [comment]
@neolit123 Thanks for your review.
It has so many fmt.Errorf in kubeadm packages,  and I'm afraid that it is a super large PR if I only submit a single PR.  According to @fabriziopandini comments, I will amend this PR to apply error.Errorf and submit them to this PR.

**Comment by yuexiao-wang** [comment]
@fabriziopandini  I have split three PR for this issue, two of which are merged. 
I find about 400 occurrences about fmt.Errorf in the rest of kubeadm package.
I have two questions:
1) we should switch to errors.Errorf when return fmt.Errorf in xxx.go and xxx_test.go, right?
2) did't switch errors.Errorf in these cases as follows:
   a) kubeadmutil.CheckErr(fmt.Errorf("failed to write the new configuration to the file %q: %v", newCfgPath, err))  
  b) notReadyDaemonSets = append(notReadyDaemonSets, fmt.Errorf("DaemonSet %q not healthy: %v", dsName, err))

**Comment by fabriziopandini** [comment]
@yuexiao-wang 
You are doing a great job. 
Small, well focused PR usually get faster review. but considering that in this case the change is really simple IMO we can relax a bit on the "small" side. Nevertheless if you are worried about the number of changes you can split the PR into smaller PRs. Up to you.

1) if I got it right, yes please switch from `return fmt.Errorf` to `return errors.New` or `return errors.Wrap` or `return errors.Errorf`
2) IMO you can switch to `errors.Wrap` in both cases as well

**Comment by neolit123** [comment]
there is also https://godoc.org/github.com/pkg/errors#Wrapf if needed in places.

**Comment by neolit123** [comment]
@yuexiao-wang 
> b) notReadyDaemonSets = append(notReadyDaemonSets, fmt.Errorf("DaemonSet %q not healthy: %v", dsName, err))

```
return fmt.Errorf("DaemonSet %q not healthy: %v", dsName, err)
```

can be changed to:
```
return errors.Wrapf(err, "DaemonSet %q not healthy", dsName)
```

so we should use Wrapf() too if needed.
otherwise you have to do:
```
return errors.Wrap(err, fmt.Sprintf("DaemonSet %q not healthy", dsName))
```

does that make sense?

**Comment by fabriziopandini** [comment]
e.g. 
`errors.New` accepts a string, but if you need to pass some argument to your string you can use `errors.Errorf("whoops: %s", "foo")` instead of `errors.New(fmt.Sprintf("whoops: %s", "foo"))`.  In other words `errors.Errorf` is a wrapper to `errors.New` that call `fmt.Sprintf`.

Same for .Wrap and .Wrapt, but if you look at source code https://github.com/pkg/errors/blob/master/errors.go everything will be clear

**Comment by yuexiao-wang** [comment]
Got it and I will look at source code

**Comment by yuexiao-wang** [comment]
CI will be green.
@fabriziopandini @neolit123 PTAL

File: cmd/kubeadm/app/phases/upgrade/policy.go [hunk 3]
```diff
@@ -52,32 +53,32 @@ func EnforceVersionPolicies(versionGetter VersionGetter, newK8sVersionStr string
 	clusterVersionStr, clusterVersion, err := versionGetter.ClusterVersion()
 	if err != nil {
 		// This case can't be forced: kubeadm has to be able to lookup cluster version for upgrades to work
-		skewErrors.Mandatory = append(skewErrors.Mandatory, fmt.Errorf("Unable to fetch cluster version: %v", err))
+		skewErrors.Mandatory = append(skewErrors.Mandatory, errors.Wrap(err, "Unable to fetch cluster version: %v"))
```
Comment (neolit123) [inline cmd/kubeadm/app/phases/upgrade/policy.go]: please remove `: %v` at the end.

File: cmd/kubeadm/app/util/marshal.go [hunk 4]
```diff
@@ -97,20 +97,20 @@ func SplitYAMLDocuments(yamlBytes []byte) (map[schema.GroupVersionKind][]byte, e
...
+			errs = append(errs, pkgerrors.Errorf("invalid configuration: kind %q is specified twice in YAML file", typeMetaInfo.Kind))
 			continue
 		}
 		knownKinds[typeMetaInfo.Kind] = true
 
 		// Build a GroupVersionKind object from the deserialized TypeMeta object
 		gv, err := schema.ParseGroupVersion(typeMetaInfo.APIVersion)
 		if err != nil {
-			errs = append(errs, fmt.Errorf("unable to parse apiVersion: %v", err))
+			errs = append(errs, pkgerrors.Wrap(err, "unable to parse apiVersion: %v"))
```
Comment (neolit123) [inline cmd/kubeadm/app/util/marshal.go]: release remove `: %v`

File: cmd/kubeadm/app/util/system/kernel_validator.go [hunk 5]
```diff
@@ -218,14 +220,14 @@ func (k *KernelValidator) getKernelConfigReader() (io.Reader, error) {
 		// config module and check again.
 		output, err := exec.Command(modprobeCmd, configsModule).CombinedOutput()
 		if err != nil {
-			return nil, fmt.Errorf("unable to load kernel module %q: output - %q, err - %v",
-				configsModule, output, err)
+			return nil, pkgerrors.Wrapf(err, "unable to load kernel module %q: output - %q, err - ",
```
Comment (neolit123) [inline cmd/kubeadm/app/util/system/kernel_validator.go]: lets change the end to `output: %q, err`

**Review by neolit123 — COMMENTED** [review]
@yuexiao-wang thanks a lot for this big cleanup! awesome.
added some minor comments.

@kubernetes/sig-cluster-lifecycle-pr-reviews 
ideally, we need more reviewers on this PR.

File: cmd/kubeadm/app/phases/certs/certs.go [hunk 6]
```diff
@@ -52,7 +54,7 @@ func CreatePKIAssets(cfg *kubeadmapi.InitConfiguration) error {
 	}
 
 	if err := certTree.CreateTree(cfg); err != nil {
-		return fmt.Errorf("Error creating PKI assets: %v", err)
+		return errors.Wrap(err, "Error creating PKI assets")
```
[the remark this answers is not in the document]
Comment (yuexiao-wang) [reply cmd/kubeadm/app/phases/certs/certs.go]: thanks
I will fix it and check all the cleanup in this PR again

File: cmd/kubeadm/app/phases/certs/pkiutil/pki_helpers.go [hunk 7]
```diff
@@ -38,12 +39,12 @@ import (
 func NewCertificateAuthority(config *certutil.Config) (*x509.Certificate, *rsa.PrivateKey, error) {
 	key, err := certutil.NewPrivateKey()
 	if err != nil {
-		return nil, nil, fmt.Errorf("unable to create private key [%v]", err)
+		return nil, nil, errors.Wrap(err, "unable to create private key [%v]")
```
[the remark this answers is not in the document]
Comment (yuexiao-wang) [reply cmd/kubeadm/app/phases/certs/pkiutil/pki_helpers.go]: got it

File: cmd/kubeadm/app/phases/certs/pkiutil/pki_helpers.go [hunk 8]
```diff
@@ -86,12 +87,12 @@ func WriteCertAndKey(pkiPath string, name string, cert *x509.Certificate, key *r
...
 func WriteCert(pkiPath, name string, cert *x509.Certificate) error {
 	if cert == nil {
-		return fmt.Errorf("certificate cannot be nil when writing to file")
+		return errors.New("certificate cannot be nil when writing to file")
 	}
 
 	certificatePath := pathForCert(pkiPath, name)
 	if err := certutil.WriteCert(certificatePath, certutil.EncodeCertPEM(cert)); err != nil {
-		return fmt.Errorf("unable to write certificate to file %q: [%v]", certificatePath, err)
+		return errors.Errorf(certificatePath, err, "unable to write certificate to file %q: [%v]")
```
[the remark this answers is not in the document]
Comment (yuexiao-wang) [reply cmd/kubeadm/app/phases/certs/pkiutil/pki_helpers.go]: got it.

File: cmd/kubeadm/app/phases/upgrade/policy.go (hunk 3 above)
Comment (yuexiao-wang) [reply cmd/kubeadm/app/phases/upgrade/policy.go]: ok

File: cmd/kubeadm/app/util/system/kernel_validator.go (hunk 5 above)
Comment (yuexiao-wang) [reply cmd/kubeadm/app/util/system/kernel_validator.go]: i want to change it to `unable to load kernel module: %q, output: %q, err`

File: cmd/kubeadm/app/util/marshal.go (hunk 4 above)
Comment (yuexiao-wang) [reply cmd/kubeadm/app/util/marshal.go]: got it

**Comment by yuexiao-wang** [comment]
I update the PR and split to two PR, which is convenient to review

**Comment by yuexiao-wang** [comment]
@fabriziopandini  I think it is ready after CI is green.

**Comment by fabriziopandini** [comment]
@yuexiao-wang if you need help feel free to contact me on slack.
For the error in pull-kubernetes-verify hack/verify-gofmt.sh and hack/update-gofmt.sh can help you

**Comment by yuexiao-wang** [comment]
@fabriziopandini  Thanks. I have run hack/verify-gofmt.sh and hack/update-gofmt.sh before submit my PR. there are not errors happen in local, but it failed in master. I don't know why

**Comment by fabriziopandini** [comment]
@yuexiao-wang In that case I suggest to rebase your PR

**Pull request merged**